Add buildTargets/jvmEnvironment endpoint

lukaszwawrzyk opened the pull request:
Depends on https://github.com/build-server-protocol/build-server-protocol/pull/113
Just implements the new endpoint, to currently be exactly the same as jvmTestEnvironment.
Fills in new optional parameters javaHome and javaVersion, currently based on the jvm that bloop is running on. In future we may allow to customize it through Config.Project

olafurpg commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,13 @@ final class BloopBspServices(
     bsp.BuildTargetIdentifier(project.bspUri)
 
+  private def toJvmBuildTarget(project: Project): bsp.JvmBuildTarget = {
+    val javaHome = sys.props.get("java.home").map(s => bsp.Uri(Paths.get(s).toUri))
```
nit: `sys.props` allocates a new immutable map on every method call so I personally use `Option(System.getProperty(..))` instead

olafurpg commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,13 @@ final class BloopBspServices(
     bsp.BuildTargetIdentifier(project.bspUri)
 
+  private def toJvmBuildTarget(project: Project): bsp.JvmBuildTarget = {
+    val javaHome = Option(System.getProperty("java.home")).map(s => bsp.Uri(Paths.get(s).toUri))
```
I just learned from an offline discussion with @jvican that Bloop supports configuring custom Java home per-project. See `.platform` here https://github.com/scalacenter/bloop/blob/906110dc384dbebb8d7d1b2f62bd226231a4f3bf/frontend/src/main/scala/bloop/data/Project.scala#L52

We should use `JdkConfig.javaHome` instead of the `java.home` system property.

olafurpg commented on build.sbt:
```diff
@@ -27,6 +27,7 @@ val benchmarkBridge = project
 lazy val bloopShared = (project in file("shared"))
   .settings(
     name := "bloop-shared",
+    resolvers += Resolver.sonatypeRepo("snapshots"),
```
We should cut a new stable BSP release so we don't use SNAPSHOTs

jvican commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,20 @@ final class BloopBspServices(
     bsp.BuildTargetIdentifier(project.bspUri)
 
+  private def toJvmBuildTarget(project: Project): bsp.JvmBuildTarget = {
+    Option(project.platform)
```
We have some of this logic already in a method called `project.jdkConfig`. What do you think about using that instead of this? We can also replace the values of `java.home` and the java version and use those that are available in `JavaRuntime`.

lukaszwawrzyk replied:
yeah, I guess it is enough to use project.jdkConfig, anyway it sets the correct jvm default config if not provided, and if it returns none, JvmBuildTarget can be empty as it means native or js.

jvican commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,20 @@ final class BloopBspServices(
+        val javaHome = Option(System.getProperty("java.home")).map(s => bsp.Uri(Paths.get(s).toUri))
+        val javaVersion = Option(System.getProperty("java.vm.specification.version"))
+        bsp.JvmBuildTarget(javaHome, javaVersion)
+      }
```
Isn't this method returning a jvm build target always, regardless of the platform?

jvican replied:
Why return a `JvmBuildTarget` here for projects that are not JVM?

jvican commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,20 @@ final class BloopBspServices(
+      }
+      .getOrElse {
+        val javaHome = Option(System.getProperty("java.home")).map(s => bsp.Uri(Paths.get(s).toUri))
+        val javaVersion = Option(System.getProperty("java.vm.specification.version"))
```
Are you really use you want `java.vm.specification.version` here instead of `java.version`?

jvican commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -863,6 +864,17 @@ final class BloopBspServices(
     bsp.BuildTargetIdentifier(project.bspUri)
 
+  private def toJvmBuildTarget(project: Project): bsp.JvmBuildTarget = {
+    project.jdkConfig
```
Nit: could we make all of this as concise as possible by using pattern matching here ?

lukaszwawrzyk replied:
ok

lukaszwawrzyk commented:
> What happened to the use of `java.specification.vm.version`?

So I addressed the comment about not filling in JvmBuildTarget in case when target is not jvm.
If project.jdkConfig is None, it means it is Native or Js, so in such case I put `JvmBuildTarget(None, None)` rather than bloop runtime jvm. I actually never fill in javaVersion, as it is not available in `project.platform`, and I can't assume that `sys.props("java.version")` is valid for given `project.jdkConfig.javaHome`

jvican commented:
@lukaszwawrzyk Great, thanks for the changes!
  
>  so in such case I put JvmBuildTarget(None, None) rather than bloop runtime jvm. 
  
Why do we return `JvmBuildTarget(None, None)` instead of `None`? This looks like a limitation in the BSP design of `JvmBuildTarget`, we should not assume every `ScalaBuildTarget` has a `JvmBuildTarget`, it should accept `Option[JvmBuildTarget]` instead.
  
@olafurpg Any thoughts?

olafurpg commented:
Good catch! The `ScalaBuildTarget.jvmBuildTarget` field needs to be typed as `Option[BuildTarget]`, otherwise new clients won't be able to parse `ScalaBuildTarget` from old servers.

olafurpg commented:
We will need to do the same change for bsp4j, this field is optional.

olafurpg commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -2,6 +2,7 @@ package bloop.bsp
 
 import java.io.InputStream
 import java.net.URI
+import java.nio.file.Paths
```
Unused import

jvican commented:
Spoke with @olafurpg offline and we agreed on merging this so that we don't need to wait for another review cycle to merge this :smile: We'll remove the unused import whenever we set up scalafix automatically in the bloop build.
